x/exchange/keeper/keeper.go

```diff
@@ -0,0 +1,270 @@
+	// an order counts as acceptance of the stuff to receive (that they defined when creating the order).
+	ctx := quarantine.WithBypass(ctxIn)
+	if len(inputs) == 1 && len(outputs) == 1 {
+		// If there's only one of each, we use SendCoins for the nicer events.
+		if !exchange.CoinsEquals(inputs[0].Coins, outputs[0].Coins) {
+			return fmt.Errorf("input coins %q does not equal output coins %q",
+				inputs[0].Coins, outputs[0].Coins)
+		}
+		fromAddr, err := sdk.AccAddressFromBech32(inputs[0].Address)
+		if err != nil {
+			return fmt.Errorf("invalid inputs[0] address %q: %w", inputs[0].Address, err)
+		}
+		toAddr, err := sdk.AccAddressFromBech32(outputs[0].Address)
+		if err != nil {
+			return fmt.Errorf("invalid outputs[0] address %q: %w", outputs[0].Address, err)
+		}
+		return k.bankKeeper.SendCoins(ctx, fromAddr, toAddr, inputs[0].Coins)
+	}
+	return k.bankKeeper.InputOutputCoins(ctx, inputs, outputs)
+}
+
+// CalculateExchangeSplit calculates the amount that the exchange will keep of the provided fee.
+func (k Keeper) CalculateExchangeSplit(ctx sdk.Context, feeAmt sdk.Coins) sdk.Coins {
+	if feeAmt.IsZero() {
+		return nil
+	}
+	exchangeAmt := make(sdk.Coins, 0, len(feeAmt))
+	for _, coin := range feeAmt {
+		if coin.Amount.IsZero() {
+			continue
+		}
+
+		split := int64(k.GetExchangeSplit(ctx, coin.Denom))
+		if split == 0 {
+			continue
+		}
+
+		splitAmt, splitRem := exchange.QuoRemInt(coin.Amount.Mul(sdkmath.NewInt(split)), TenKInt)
+		if !splitRem.IsZero() {
+			splitAmt = splitAmt.Add(OneInt)
+		}
+		exchangeAmt = append(exchangeAmt, sdk.NewCoin(coin.Denom, splitAmt))
+	}
+	if exchangeAmt.IsZero() {
+		return nil
+	}
+	return exchangeAmt
+}
+
+// CollectFee will transfer the fee amount to the market account,
+// then the exchange's cut from the market to the fee collector.
+// If you have fees to collect from multiple payers, consider using CollectFees.
+func (k Keeper) CollectFee(ctx sdk.Context, marketID uint32, payer sdk.AccAddress, fee sdk.Coins) error {
+	if fee.IsZero() {
+		return nil
+	}
+	exchangeSplit := k.CalculateExchangeSplit(ctx, fee)
+
+	marketAddr := exchange.GetMarketAddress(marketID)
+	if err := k.bankKeeper.SendCoins(ctx, payer, marketAddr, fee); err != nil {
+		return fmt.Errorf("error transferring %s from %s to market %d: %w", fee, payer, marketID, err)
+	}
+	if !exchangeSplit.IsZero() {
+		if err := k.bankKeeper.SendCoinsFromAccountToModule(ctx, marketAddr, k.feeCollectorName, exchangeSplit); err != nil {
+			return fmt.Errorf("error collecting exchange fee %s (based off %s) from market %d: %w", exchangeSplit, fee, marketID, err)
+		}
+	}
+
+	return nil
+}
+
+// CollectFees will transfer the inputs to the market account,
+// then the exchange's cut from the market to the fee collector.
+// If there is only one input, CollectFee is used.
+func (k Keeper) CollectFees(ctx sdk.Context, marketID uint32, inputs []banktypes.Input) error {
+	if len(inputs) == 0 {
+		return nil
+	}
+	if len(inputs) == 1 {
```

Comment:
Is the only difference here the type of events? It might make it easier for listeners if they are guaranteed to get the same type of events, and it would be nice to remove this if it's redundant.

Reply:
Yes, it's events. And using `InputOutputCoins` for a 1-input 1-output transfer has caused confusion in the past, mostly because `MultiSend` (or `InputOutputCoins`) is almost never used. People are more used to the `SendCoins` events, and I expect the norm here to be 1-to-1. If I ever update `InputOutputCoins` to again allow many to many, it'll actually clean up a whole lot of logic, and I'd always use it since there'd always be at least two inputs and outputs; the tricky thing there is the `SendRestriction`s.